api/envoy/service/discovery/v3/discovery.proto

```diff
@@ -199,6 +205,11 @@ message DeltaDiscoveryRequest {
   // failed to update configuration. The *message* field in *error_details*
   // provides the Envoy internal exception related to the failure.
   google.rpc.Status error_detail = 7;
+
+  // Constraints on the set of cache keys for resources of type *type_url*.
+  // Populated only on the first request for a given resource type and
+  // when the constraints change.
```

Comment:
+1 on this comment. This is essentially the big gap we have today in terms of how `Node` is treated. We need to be very clear on what the "effective cache key is" at all times and how this relates to new/stale/updated subscriptions.

Reply:
The intent here is that constraints apply to the *resource type*, not to individual resources.

The assumption here is that a client will only be using one set of cache key constraints for a given resource type at any given time.  The constraints may change over time, but when they do, that doesn't affect the actual set of subscriptions; it just changes what resource contents the server is sending for the subscribed resources.

For example, let's say that a server has two variants of a resource named `foo`, one with cache key `shard=0` and another with cache key `shard=1`.  We could have the following sequence of events:

1. The client initially sends a request with `resource_names_subscribe` set to `foo` and `cache_key_constraints` set to `shard=0`.  The server sends back the resource `foo` with contents from the resource with cache key `shard=0`.
2. Then, if the client is told (via some out-of-band mechanism) to change to shard 1, it will send another request with `resource_names_subscribe` unset (because it's not actually changing the set of resources it's subscribed to) but with `cache_key_constraints` set to `shard=1`.  The server will react by sending an update for resource `foo` that has the contents from the resource with cache key `shard=1`.  From the client's perspective, this looks like an update for the resource that it already subscribed to, not like a different resource.

Note that for the incremental transport protocol, this means that the authoritative server should assign a unique per-resource version to each variant of the resource, so that the client has a way to ACK/NACK when this change occurs.

Reply:
It seems that there's an underlying assumption, that when a client modifies the `cache_key_constraints`, all the previous resources of the same type (with the previous keys) should be discarded, and the server shouldn't send any updates on them.
That said, this might be dependent on the semantics of the modified constraint, and what was the change. For example:
1. The client sends a request with `cache_key_constraints` set to `shard_range=[0,2]`. The server sends back the resources for that range.
2. The client then wishes to modify the range and sends a `cache_key_constraints` set to `shard_range=[0,5]`. In this case the server could potentially understand that the client already knows about the resources that are in range [0,2] and only add more resource, and the client needs to be aware of them (also both the client and server should agree what is the desired behavior when ranges partially overlap).

Actually, it might depend on the semantics of the range parameter, as it can imply returning all resources in that range (as in the example above), or just one of the resources that is in the range (thinking about the minor-version use-case).

Something similar may happen for an optional entry in the `cache_key_constraints` that becomes non-optional in a future request. Should the client discard all the previous resources, or keep the ones that already match the new (non-optional) constraint.

The simplest option would probably be assuming no previous resources should be used, but then there is more noise, and repeated messages will be sent.
Another option is to make this explicit per-constraint.
WDYT?

Reply:
> It seems that there's an underlying assumption, that when a client modifies the `cache_key_constraints`, all the previous resources of the same type (with the previous keys) should be discarded, and the server shouldn't send any updates on them.

No, the client should not discard anything.

The intent here is that a request that changes the `cache_key_constraints` *does not have any effect at all on the set of resource locators that the client is subscribed to*.  The client is still subscribed to the same set of resources it was subscribed to before it sent the new `cache_key_constraints`, and it should keep using the same cached values it had before.  The only meaning of this request is to tell the server that the client is changing the keys that the server should use to select the contents of the resources.

On the server side, if changing the cache key constraints causes the server to change the resource contents or the set of resources it's sending to the client, then it can send those updates explicitly.  From the client's perspective, those updates will be handled exactly the same way that they would have been if the client hadn't sent a request updating its cache key constraints.

In your example, we would be using a glob collection or a wildcard subscription, not a singleton subscription.  Using a glob collection example, it would work like this:

1. The client subscribes to `xdstp://xds.example.com/envoy.config.v3.ListenerCollection/listeners/*` with `cache_key_contraints` set to `shard=[0,2]`.
2. The server sends back resources `xdstp://xds.example.com/envoy.config.v3.ListenerCollection/listeners/shard0`, `xdstp://xds.example.com/envoy.config.v3.ListenerCollection/listeners/shard1`, and `xdstp://xds.example.com/envoy.config.v3.ListenerCollection/listeners/shard2`.
3. The client sends a new request that does not change its subscription but changes its `cache_key_constraints`  to `shard=[1,5]` (i.e., removing shard 0 and adding shards 4 and 5).
4. The server sends a deletion for `xdstp://xds.example.com/envoy.config.v3.ListenerCollection/listeners/shard0`, since that is no longer relevant to the client, and it sends an addition for `xdstp://xds.example.com/envoy.config.v3.ListenerCollection/listeners/shard4` and `xdstp://xds.example.com/envoy.config.v3.ListenerCollection/listeners/shard5`.

In other words, the client does not need to invalidate anything in its cache; it can just send the new cache key constraints, and the server will tell it what changes as a result.

>  Actually, it might depend on the semantics of the range parameter, as it can imply returning all resources in that range (as in the example above), or just one of the resources that is in the range (thinking about the minor-version use-case).

The semantics here are function of whether the client is subscribing to a collection or a singleton.  If it's a collection (or a wildcard subscription, prior to the new naming scheme), then a range means "all resources in this range".  If it's a singleton subscription, then the server needs to return exactly one matching resource, just like it does today; it can choose any of the resources in the range, but it cannot choose more than one of them.

So in the minor-version use-case, the server can pick any supported version, but it can return only one of them.  But the sharded-cluster use-case would return all resources in the range, as shown in the example above.

Reply:
I realized that this approach won't work.  It's ideal for "leaf" clients, because any individual leaf client will need only one variant of any given resource.  However, it won't work for caching xDS proxies, which need to be able to fetch and cache multiple variants of any given resource.

So I think I'm going to need to re-work this such that the cache keys are actually part of the subscription mechanism.  This unfortunately has a couple of effects:
- If a cache key changes but the underlying resource contents do not, the server will need to re-send the resource, because the change is being modeled as an unsubscribe and a re-subscribe.  (I am thinking about ways that I might be able to ameliorate this.)
- We will once again be back in the situation where we need to deal with ambiguity at the transport protocol level in terms of matching up resources with requests.  I am going to have to bite the bullet and define a crisp set of precedence rules for this.

I'll put out a new proposal once I've thought through some of these details.